File: mcp-llm-server/src/mcp_llm_server/domains/twentyq/prompts.py
"""Twenty Questions prompt loader and manager."""

import logging
from functools import lru_cache
from pathlib import Path

from mcp_llm_server.utils.prompts import load_yaml_mapping


log = logging.getLogger(__name__)

PROMPTS_DIR = Path(__file__).parent / "prompts"


def _escape_braces_for_langchain(
    text: str, keep_placeholders: list[str] | None = None
) -> str:
    """DEPRECATED: No longer needed - YAML files now use escaped braces.

    Kept for backward compatibility but simply returns text unchanged.
    JSON examples in YAML use {{ }} which LangChain treats as literal braces.
    """
    # 이제 YAML 파일에서 직접 {{ }} 사용하므로 이스케이프 불필요
    return text


@lru_cache(maxsize=16)
def _load_prompt_file(name: str) -> dict[str, str]:
    """Load a prompt YAML file."""
    path = PROMPTS_DIR / f"{name}.yml"

    log.debug("PROMPT_LOADED name=%s", name)
    return load_yaml_mapping(path)


class TwentyQPrompts:
    """Prompt templates for Twenty Questions game.

    Uses LangChain PromptTemplate for variable substitution.
    YAML files use {{ }} for literal braces in JSON examples.
    """

    # hint generation
    @staticmethod
    def hints_system(category: str | None = None) -> str:
        """Get system prompt for hint generation."""
        from langchain_core.prompts import PromptTemplate

        data = _load_prompt_file("hints")
        system = data["system"]

        if category:
            restriction = data.get("category_restriction", "")
            if restriction:
                forbidden = _get_forbidden_words(category)
                # LangChain PromptTemplate으로 변수 치환
                template = PromptTemplate.from_template(
                    restriction,
                    template_format="f-string",
                )
                restriction = template.format(
                    selectedCategory=category,
                    forbiddenWords=", ".join(forbidden),
                )
                system = system + "\n\n" + restriction

        return system

    @staticmethod
    def hints_user(secret: str) -> str:
        """Get user prompt for hint generation."""
        from langchain_core.prompts import PromptTemplate

        data = _load_prompt_file("hints")
        template = PromptTemplate.from_template(
            data["user"], template_format="f-string"
        )
        return template.format(toon=secret)

    # answer (yes/no scale)
    @staticmethod
    def answer_system() -> str:
        """Get system prompt for answering questions."""
        data = _load_prompt_file("answer")
        return data["system"]

    @staticmethod
    def answer_user(secret: str, question: str, history: str = "") -> str:
        """Get user prompt for answering questions."""
        from langchain_core.prompts import PromptTemplate

        data = _load_prompt_file("answer")
        template = PromptTemplate.from_template(
            data["user"], template_format="f-string"
        )
        result = template.format(toon=secret, question=question)
        if history:
            result = history + "\n\n" + result
        return result

    # verify guess
    @staticmethod
    def verify_system() -> str:
        """Get system prompt for verifying guesses."""
        data = _load_prompt_file("verify-answer")
        return data["system"]

    @staticmethod
    def verify_user(target: str, guess: str) -> str:
        """Get user prompt for verifying guesses."""
        from langchain_core.prompts import PromptTemplate

        data = _load_prompt_file("verify-answer")
        template = PromptTemplate.from_template(
            data["user"], template_format="f-string"
        )
        return template.format(target=target, guess=guess)

    # normalize question
    @staticmethod
    def normalize_system() -> str:
        """Get system prompt for normalizing questions."""
        data = _load_prompt_file("normalize")
        return data["system"]

    @staticmethod
    def normalize_user(question: str) -> str:
        """Get user prompt for normalizing questions."""
        from langchain_core.prompts import PromptTemplate

        data = _load_prompt_file("normalize")
        template = PromptTemplate.from_template(
            data["user"], template_format="f-string"
        )
        return template.format(question=question)

    # synonym check
    @staticmethod
    def synonym_system() -> str:
        """Get system prompt for synonym checking."""
        data = _load_prompt_file("synonym-check")
        return data["system"]

    @staticmethod
    def synonym_user(target: str, guess: str) -> str:
        """Get user prompt for synonym checking."""
        from langchain_core.prompts import PromptTemplate

        data = _load_prompt_file("synonym-check")
        template = PromptTemplate.from_template(
            data["user"], template_format="f-string"
        )
        return template.format(target=target, guess=guess)


def _get_forbidden_words(category: str) -> list[str]:
    """Get forbidden words for a category."""
    # 카테고리별 금지 단어 (정답 직접 노출 방지)
    category_forbidden: dict[str, list[str]] = {
        "음식": ["음식", "먹을 것", "식품"],
        "동물": ["동물", "생물", "생명체"],
        "사물": ["사물", "물건", "도구"],
        "장소": ["장소", "곳", "위치"],
        "인물": ["인물", "사람", "인간"],
        "개념": ["개념", "추상적", "관념"],
    }
    return category_forbidden.get(category, [category])


def get_twentyq_prompts() -> TwentyQPrompts:
    """Get the prompt manager instance."""
    return TwentyQPrompts()

File: mcp-llm-server-go/internal/domain/twentyq/prompts.go
package twentyq

import (
	"embed"
	"fmt"
	"strings"

	"github.com/park285/llm-kakao-bots/mcp-llm-server-go/internal/prompt"
)

//go:embed prompts/*.yml
var promptsFS embed.FS

// Prompts: TwentyQ 프롬프트 모음입니다.
type Prompts struct {
	prompts map[string]map[string]string
}

// NewPrompts: TwentyQ 프롬프트를 로드합니다.
func NewPrompts() (*Prompts, error) {
	loaded, err := prompt.LoadYAMLDir(promptsFS, "prompts")
	if err != nil {
		return nil, fmt.Errorf("load twentyq prompts: %w", err)
	}
	return &Prompts{prompts: loaded}, nil
}

// HintsSystem: 힌트 시스템 프롬프트를 반환합니다.
func (p *Prompts) HintsSystem(category string) (string, error) {
	data, err := p.getPrompt("hints")
	if err != nil {
		return "", err
	}
	system, err := promptField(data, "system", "hints.system")
	if err != nil {
		return "", err
	}
	if category == "" {
		return system, nil
	}

	restriction, ok := data["category_restriction"]
	if !ok || restriction == "" {
		return system, nil
	}

	forbidden := strings.Join(getForbiddenWords(category), ", ")
	formatted, err := prompt.FormatTemplate(restriction, map[string]string{
		"selectedCategory": category,
		"forbiddenWords":   forbidden,
	})
	if err != nil {
		return "", fmt.Errorf("format hints restriction: %w", err)
	}

	return system + "\n\n" + formatted, nil
}

// HintsUser: 힌트 유저 프롬프트를 반환합니다.
func (p *Prompts) HintsUser(secret string) (string, error) {
	data, err := p.getPrompt("hints")
	if err != nil {
		return "", err
	}
	template, err := promptField(data, "user", "hints.user")
	if err != nil {
		return "", err
	}
	formatted, err := prompt.FormatTemplate(template, map[string]string{
		"toon": prompt.WrapXML("secret", secret),
	})
	if err != nil {
		return "", fmt.Errorf("format hints.user: %w", err)
	}
	return formatted, nil
}

// AnswerSystem: 답변 시스템 프롬프트를 반환합니다.
func (p *Prompts) AnswerSystem() (string, error) {
	data, err := p.getPrompt("answer")
	if err != nil {
		return "", err
	}
	return promptField(data, "system", "answer.system")
}

// AnswerSystemWithSecret: Secret 정보를 포함한 시스템 프롬프트를 반환합니다.
// 암시적 캐싱 최적화: Secret 정보를 System Prompt에 통합하여 Static Prefix를 확보합니다.
// Toon 포맷이 이미 구조화되어 있으므로 XML 래핑 불필요
func (p *Prompts) AnswerSystemWithSecret(secret string) (string, error) {
	base, err := p.AnswerSystem()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s\n\n[이번 게임의 정답]\n%s\n(이 내용은 사용자에게 절대 직접 노출하지 마시오.)",
		base, secret), nil // XML 래핑 제거
}

// AnswerUser: 답변 유저 프롬프트를 반환합니다.
// 암시적 캐싱 최적화: 현재 질문만 포함하여 Cache Miss 영역을 최소화합니다.
func (p *Prompts) AnswerUser(question string) (string, error) {
	data, err := p.getPrompt("answer")
	if err != nil {
		return "", err
	}
	template, err := promptField(data, "user", "answer.user")
	if err != nil {
		return "", err
	}
	result, err := prompt.FormatTemplate(template, map[string]string{
		"question": prompt.WrapXML("question", question),
	})
	if err != nil {
		return "", fmt.Errorf("format answer.user: %w", err)
	}
	return result, nil
}

// VerifySystem: 검증 시스템 프롬프트를 반환합니다.
func (p *Prompts) VerifySystem() (string, error) {
	data, err := p.getPrompt("verify-answer")
	if err != nil {
		return "", err
	}
	return promptField(data, "system", "verify-answer.system")
}

// VerifyUser: 검증 유저 프롬프트를 반환합니다.
func (p *Prompts) VerifyUser(target string, guess string) (string, error) {
	data, err := p.getPrompt("verify-answer")
	if err != nil {
		return "", err
	}
	template, err := promptField(data, "user", "verify-answer.user")
	if err != nil {
		return "", err
	}
	formatted, err := prompt.FormatTemplate(template, map[string]string{
		"target": prompt.WrapXML("target", target),
		"guess":  prompt.WrapXML("guess", guess),
	})
	if err != nil {
		return "", fmt.Errorf("format verify-answer.user: %w", err)
	}
	return formatted, nil
}

// NormalizeSystem: 정규화 시스템 프롬프트를 반환합니다.
func (p *Prompts) NormalizeSystem() (string, error) {
	data, err := p.getPrompt("normalize")
	if err != nil {
		return "", err
	}
	return promptField(data, "system", "normalize.system")
}

// NormalizeUser: 정규화 유저 프롬프트를 반환합니다.
func (p *Prompts) NormalizeUser(question string) (string, error) {
	data, err := p.getPrompt("normalize")
	if err != nil {
		return "", err
	}
	template, err := promptField(data, "user", "normalize.user")
	if err != nil {
		return "", err
	}
	formatted, err := prompt.FormatTemplate(template, map[string]string{
		"question": prompt.WrapXML("question", question),
	})
	if err != nil {
		return "", fmt.Errorf("format normalize.user: %w", err)
	}
	return formatted, nil
}

// SynonymSystem: 유사어 시스템 프롬프트를 반환합니다.
func (p *Prompts) SynonymSystem() (string, error) {
	data, err := p.getPrompt("synonym-check")
	if err != nil {
		return "", err
	}
	return promptField(data, "system", "synonym-check.system")
}

// SynonymUser: 유사어 유저 프롬프트를 반환합니다.
func (p *Prompts) SynonymUser(target string, guess string) (string, error) {
	data, err := p.getPrompt("synonym-check")
	if err != nil {
		return "", err
	}
	template, err := promptField(data, "user", "synonym-check.user")
	if err != nil {
		return "", err
	}
	formatted, err := prompt.FormatTemplate(template, map[string]string{
		"target": prompt.WrapXML("target", target),
		"guess":  prompt.WrapXML("guess", guess),
	})
	if err != nil {
		return "", fmt.Errorf("format synonym-check.user: %w", err)
	}
	return formatted, nil
}

func (p *Prompts) getPrompt(name string) (map[string]string, error) {
	if p == nil {
		return nil, fmt.Errorf("twentyq prompts not initialized")
	}
	promptMap, err := prompt.Get(p.prompts, name, "twentyq")
	if err != nil {
		return nil, fmt.Errorf("get twentyq prompt %s: %w", name, err)
	}
	return promptMap, nil
}

func promptField(data map[string]string, key string, label string) (string, error) {
	value, err := prompt.Field(data, key, label)
	if err != nil {
		return "", fmt.Errorf("get twentyq prompt field %s: %w", label, err)
	}
	return value, nil
}

func getForbiddenWords(category string) []string {
	categoryForbidden := map[string][]string{
		"음식": {"음식", "먹을 것", "식품"},
		"동물": {"동물", "생물", "생명체"},
		"사물": {"사물", "물건", "도구"},
		"장소": {"장소", "곳", "위치"},
		"인물": {"인물", "사람", "인간"},
		"개념": {"개념", "추상적", "관념"},
	}
	if forbidden, ok := categoryForbidden[category]; ok {
		return forbidden
	}
	return []string{category}
}
